lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+        LLVM_DEBUG(llvm::dbgs()
+                   << "hop-injector: rejecting bb" << block.getDebugID()
+                   << " b/c no non-load uses.\n");
+        continue; // could not be an initializing block.
+      }
+
+      // Determine if this `block` is initializing, that is:
+      //
+      //     InAvailability ≡ merge(OutAvailability(predecessors(block)))
+      //                    ≠ Yes
+      //               AND
+      //     OutAvailability(block)) = Yes
+      //
+      // For a block with no successors, we assume it's out-avail is Yes,
+      // and a block with no predecessors has in-avail of non-Yes.
+
+      auto outSet = info.OutAvailability;
+      if (block.getNumSuccessors() != 0 && !outSet.isAllYes()) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "hop-injector: rejecting bb" << block.getDebugID()
+                   << " b/c non-Yes OUT avail\n");
+        continue; // then this block never sees TheMemory initialized.
+      }
+
+      AvailabilitySet inSet(outSet.size());
+      auto const &predecessors = block.getPredecessorBlocks();
+      for (const auto &pred : predecessors)
+        inSet.mergeIn(getBlockInfo(pred).OutAvailability);
+
+      if (inSet.isAllYes()) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "hop-injector: rejecting bb" << block.getDebugID()
+                   << " b/c all-Yes IN avail\n");
+        continue; // then this block always sees TheMemory initialized.
+      }
+
+      initializingBlocks.push_back({&block, inSet});
+
+      LLVM_DEBUG(llvm::dbgs()
+                 << "hop-injector: bb" << block.getDebugID()
+                 << " is initializing block with in-availability: ");
+      LLVM_DEBUG(inSet.dump(llvm::dbgs()));
+      LLVM_DEBUG(llvm::dbgs() << "\n");
+    } // end loop
+  }
+
+  // Step 2: Scan each block to find the first non-load use that
+  // fully-initializes TheMemory.
+
+  // Tracks status of each element of TheMemory as we scan through the block.
+  AvailabilitySet localAvail;
+  for (auto const &entry : initializingBlocks) {
+    auto *block = entry.first;
+
+    // set-up initial availability at the block's entry-point.
+    localAvail = entry.second;
+
+    auto bbi = block->begin(); // our cursor and eventual insertion-point.
+    for (auto bbe = block->end(); bbi != bbe; ++bbi) {
+      auto *inst = &*bbi;
+
+      auto result = NonLoadUses.find(inst);
+      if (result == NonLoadUses.end())
+        continue; // not a possible store
+
+      // Mark the tuple elements involved in this use as defined.
+      for (unsigned use : result->second) {
+        auto const &instUse = Uses[use];
+        for (unsigned i = instUse.FirstElement;
+             i < instUse.FirstElement + instUse.NumElements; ++i)
+          localAvail.set(i, DIKind::Yes);
+      }
+
+      // Have we found the instruction that initializes TheMemory?
+      if (localAvail.isAllYes()) {
+        bbi++; // bump to next instruction for an insertion point.
```

Comment:
Is it always guaranteed that `inst` is _not_ a terminator instruction? E.g. can `use` be an inout argument of a `try_apply`?
Because if it's a terminator instruction, you cannot insert the `hop_to_executor` after it.
Instead you can use `SILBuilderWithScope::insertAfter`.

Reply:
I decided to use `insertAfter` but also added an assert ensuring its not a terminator, because `insertAfter` technically doesn't support some kinds of successors (i.e., those with multiple predecessors).

It should be always guaranteed because I don't see any kinds of terminators that do a store, and `inout` uses are banned before the stored property is initialized.

Reply:
Good point though thank you for catching this.